clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -675,36 +676,41 @@ private ListOffsetResult fetchOffsetsByTimes(Map<TopicPartition, Long> timestamp
-                }
+            if (position != null) {
+                if (completedFetch.nextFetchOffset == position.offset) {
+                    List<ConsumerRecord<K, V>> partRecords = completedFetch.fetchRecords(maxRecords);
+
+                    log.trace("Returning {} fetched records at offset {} for assigned partition {}",
+                            partRecords.size(), position, completedFetch.partition);
+
+                    if (completedFetch.nextFetchOffset > position.offset) {
+                        FetchPosition nextPosition = new FetchPosition(
+                                completedFetch.nextFetchOffset,
+                                completedFetch.lastEpoch,
+                                position.currentLeader);
+                        log.trace("Update fetching position to {} for partition {}", nextPosition, completedFetch.partition);
+                        subscriptions.position(completedFetch.partition, nextPosition);
+                    }
 
-                Long partitionLag = subscriptions.partitionLag(completedFetch.partition, isolationLevel);
-                if (partitionLag != null)
-                    this.sensors.recordPartitionLag(completedFetch.partition, partitionLag);
+                    Long partitionLag = subscriptions.partitionLag(completedFetch.partition, isolationLevel);
+                    if (partitionLag != null)
+                        this.sensors.recordPartitionLag(completedFetch.partition, partitionLag);
 
-                Long lead = subscriptions.partitionLead(completedFetch.partition);
-                if (lead != null) {
-                    this.sensors.recordPartitionLead(completedFetch.partition, lead);
-                }
+                    Long lead = subscriptions.partitionLead(completedFetch.partition);
+                    if (lead != null) {
+                        this.sensors.recordPartitionLead(completedFetch.partition, lead);
+                    }
 
-                return partRecords;
+                    return partRecords;
+                } else {
+                    // these records aren't next in line based on the last consumed position, ignore them
+                    // they must be from an obsolete request
+                    log.debug("Ignoring fetched records for {} at offset {} since the current position is {}",
+                            completedFetch.partition, completedFetch.nextFetchOffset, position);
+                }
             } else {
-                // these records aren't next in line based on the last consumed position, ignore them
-                // they must be from an obsolete request
-                log.debug("Ignoring fetched records for {} at offset {} since the current position is {}",
-                        completedFetch.partition, completedFetch.nextFetchOffset, position);
+                log.warn("Ignoring fetched records for {} at offset {} since the current position is undefined",
```

Comment:
This comment applies to a few of the added null checks where we have already validated that the partition is "fetchable." I am wondering if it would be more consistent to raise an exception.

Reply:
Would IllegalStateException make sense here (since we just checked that the partition was fetchable)?

Reply:
Sounds fine to me.